### PR TITLE
printing - fixes issue 21814 and now some expressions are printed correctly

### LINKED ISSUE #21814: Printing of unevaluated Mul needs brackets
The second version here should use brackets or something:
```python
In [16]: Mul(Rational(1, 2), x + y, evaluate=False)
Out[16]: 
x + y
─────
  2  

In [17]: Mul(x + y, Rational(1, 2), evaluate=False)
Out[17]: x + y⋅1/2
```

### PR DESCRIPTION
**Pull request by praneethratna**
#### Brief description of what is fixed or changed
Fixes #21814 and the now the Mul expression Mul(x + y, Rational(1, 2), evaluate=False) is printed as (x+y)*2 also fixed the printing in pretty, upretty and latex.

#### Release Notes

* printing
     * Fixed printing of unevaluated Mul which now outputs with brackets

### CODE REVIEW

**Comment by praneethratna** [comment]
@oscarbenjamin Sorry I've made a mess with commit so i opened a new PR , I've also added the tests mentioned by @namannimmo10 could you please review?
Thanks!

File: sympy/printing/pretty/tests/test_pretty.py [hunk 1]
```diff
@@ -996,6 +996,31 @@ def test_negative_fractions():
...
+    assert pretty(expr) == "1*(x + y)"
+    assert upretty(expr) == "1⋅(x + y)"
+    expr = Mul(x - y, S.One, evaluate=False)
+    assert pretty(expr) == "x - y"
+    assert upretty(expr) == "x - y"
+    expr = Mul(Rational(1, 2), x - y, S.One, x + y, evaluate=False)
+    assert pretty(expr) == "1/2*(x - y)*1*(x + y)"
+    assert upretty(expr) == "1/2⋅(x - y)⋅1⋅(x + y)"
+    expr = Mul(x + y, Rational(3, 4), S.One, y - z, evaluate=False)
+    assert pretty(expr) == "3*(x + y)*(y - z)\n-----------------\n        4        "
```
Comment (namannimmo10) [inline sympy/printing/pretty/tests/test_pretty.py:1015]: I don't see how https://github.com/sympy/sympy/pull/21819#issuecomment-894163417 is true.

Comment (praneethratna) [reply sympy/printing/pretty/tests/test_pretty.py:1015]: @namannimmo10  On this branch , In my local setup
```
>>> import sympy
>>> from sympy import *
>>> x,y,z = symbols('x y z')
>>> Mul(x + y, Rational(3, 4), S.One, y - z, evaluate=False)
```
outputs `(x + y)⋅(3/4)⋅1⋅(y - z)`

File: sympy/printing/latex.py [hunk 2]
```diff
@@ -546,7 +546,7 @@ def convert_args(args):
         # XXX: _print_Pow calls this routine with instances of Pow...
         if isinstance(expr, Mul):
             args = expr.args
-            if args[0] is S.One or any(isinstance(arg, Number) for arg in args[1:]):
+            if args[0] is S.One or any(isinstance(arg, Number) for arg in args[1:]) and args[0].is_number:
```
Comment (oscarbenjamin) [inline sympy/printing/latex.py:549]: I don't think that `is_number` is the right thing to check here. Why is this needed?

The `is_number` predicate is for checking whether or not an expression can be evaluated using `evalf` to give an explicit approximate complex floating point value. I don't see how that can be the appropriate property to check here.

Comment (praneethratna) [reply sympy/printing/latex.py:549]: In the current master,
```
>>>latex(Mul(x + y, Rational(1, 2), evaluate=False))
'\\left(x + y\\right) \\frac{1}{2}'
>>> latex(Mul(Rational(1, 2), x + y, evaluate=False))
'\\frac{x + y}{2}'
```
After this change , in this branch,
```
>>>latex(Mul(x + y, Rational(1, 2), evaluate=False))
'\\frac{x + y}{2}'
>>> latex(Mul(Rational(1, 2), x + y, evaluate=False))
'\\frac{x + y}{2}'
```
 But I'm also not sure whether the is_number is plausible or not.

Comment (oscarbenjamin) [reply sympy/printing/latex.py:549]: These two should not display the same:

> >>>latex(Mul(x + y, Rational(1, 2), evaluate=False))
> '\\frac{x + y}{2}'
> >>> latex(Mul(Rational(1, 2), x + y, evaluate=False))
> '\\frac{x + y}{2}'

It is deliberate that unevaluated expressions should look different from evaluated expressions. In this case the current master output is correct.

Comment (praneethratna) [reply sympy/printing/latex.py:549]: But here aren't they both unevaluated?

Comment (praneethratna) [reply sympy/printing/latex.py:549]: > These two should not display the same:
> 
> > > > > latex(Mul(x + y, Rational(1, 2), evaluate=False))
> > > > > '\frac{x + y}{2}'
> > > > > latex(Mul(Rational(1, 2), x + y, evaluate=False))
> > > > > '\frac{x + y}{2}'
> 
> It is deliberate that unevaluated expressions should look different from evaluated expressions. In this case the current master output is correct.
But here aren't they both unevaluated?

Comment (oscarbenjamin) [reply sympy/printing/latex.py:549]: They are both unevaluated but the first has the Rational as first arg. The `if` check that you are disabling is there deliberately so that expressions with non-canonical args do not print the same as expressions with canonical args.

Comment (praneethratna) [reply sympy/printing/latex.py:549]: Yes , I understood it now. I'll try some other way of fixing issue #21814

Comment (praneethratna) [reply sympy/printing/latex.py:549]: @oscarbenjamin I think i found a fix but I've a query:
does `Mul(x - y, S.One, evaluate=False)` should output just `x - y` or `(x - y)*1` , according to my fix it will output the latter is it the correct one?

Comment (oscarbenjamin) [reply sympy/printing/latex.py:549]: Here `(x - y)*1` is correct.

File: sympy/printing/pretty/tests/test_pretty.py (hunk 1 above)
Comment (praneethratna) [reply sympy/printing/pretty/tests/test_pretty.py]: @namannimmo10 This is fixed now ,
`>>> pretty(Mul(x + y, Rational(3, 4), S.One, y - z, evaluate=False))`
outputs `(x + y)*1*3/4*5/6`  and
`>>> upretty(Mul(x + y, Rational(3, 4), S.One, y - z, evaluate=False))`
outputs `(x + y)⋅1⋅3/4⋅5/6`

File: sympy/printing/pretty/stringpict.py [hunk 3]
```diff
@@ -441,8 +441,8 @@ def __mul__(self, *others):
         if len(others) == 0:
             return self # We aren't actually multiplying... So nothing to do here.
         args = self
-        if args.binding > prettyForm.MUL:
-            arg = stringPict(*args.parens())
+        if args.binding > prettyForm.MUL and len(args[0].split()) > 1:
+            args = stringPict(*args.parens())
```
Comment (oscarbenjamin) [inline sympy/printing/pretty/stringpict.py:445]: What exactly is `args[0]` here? Is it a string?

Is this basically checking for spaces and then adding brackets if there are any?

Comment (praneethratna) [reply sympy/printing/pretty/stringpict.py:445]: Yeah , Here args[0] is a string and the first one in multiplication part , so Basically I'm checking whether it is a expression or not because when there is something like `-A*1/2* we aren't adding any brackets to `A` but when we have something like 
`x + y + z*1/2` we need to add brackets to `x + y + z`.

Comment (oscarbenjamin) [reply sympy/printing/pretty/stringpict.py:445]: I think that there should be a more robust way of checking this than looking for spaces in the string. What if the string just had some spaces in it for a different reason?

I don't know the printing code so well though so I'm not sure what's best.

CC @oscargus

Comment (oscargus) [reply sympy/printing/pretty/stringpict.py:445]: I don't know the pretty printer that well, but I also think that it is a bit doubtful that it will always work. (Also, I think that `len(args[0].split()) > 1` should be replaced by `args[0].count(' ')` if this method is to be used.)

I think one should add a load of test cases if one considers this approach, including multiplication of more complicated functions that will print on multiple lines etc. What about (sin(x)/cos(x))*(1/2)? I'm also a bit surprised that the rational number is not printed as a fraction.

By the way, if I understand correctly, the existing code detects that it is multiplied by a fraction and then realizes that there is not need to add brackets since it will be written out as a fraction and you are now trying to add the brackets again. It would make more sense to approach it before it realizes that there shouldn't be any fractions, just as if it was `(x + y + z)*q`. With that said, I do not fully understand how it works.

Personally, I do not fully see the fuzz with the non-evaluated printing stuff. Much effort has been put into making the expressions print nicely, so just trying to bypass that doesn't really seem to make sense to me. But that is another question.

File: sympy/printing/pretty/tests/test_pretty.py [hunk 4]
```diff
@@ -996,6 +996,30 @@ def test_negative_fractions():
...
+    assert pretty(expr) == "(x - y)*1"
+    assert upretty(expr) == "(x - y)⋅1"
+    expr = Mul(Rational(1, 2), x - y, S.One, x + y, evaluate=False)
+    assert pretty(expr) == "1/2*(x - y)*1*(x + y)"
+    assert upretty(expr) == "1/2⋅(x - y)⋅1⋅(x + y)"
+    expr = Mul(x + y, Rational(3, 4), S.One, y - z, evaluate=False)
+    assert pretty(expr) == "(x + y)*3/4*1*(y - z)"
+    assert upretty(expr) == "(x + y)⋅3/4⋅1⋅(y - z)"
+    expr = Mul(x + y, Rational(1, 1), Rational(3, 4), Rational(5, 6),evaluate=False)
+    assert pretty(expr) == "(x + y)*1*3/4*5/6"
```
Comment (oscargus) [inline sympy/printing/pretty/tests/test_pretty.py:1018]: Shouldn't `Rational(1,1)`, strictly, print as `1/1`? What about `Rational(2,4)`? Or is that not kept with `evaluate=False`?

File: sympy/printing/pretty/stringpict.py (hunk 3 above)
Comment (oscargus) [reply sympy/printing/pretty/stringpict.py:445]: Also, why is `arg` changed to `args`? It almost seems like the code wasn't working before, since `arg` was not used...

Comment (praneethratna) [reply sympy/printing/pretty/stringpict.py:445]: Before , it was there as we should'nt be adding brackets to `A` whenever we multiply something like `-A*B*C` .

File: sympy/printing/pretty/tests/test_pretty.py (hunk 4 above)
Comment (praneethratna) [reply sympy/printing/pretty/tests/test_pretty.py:1018]: @oscargus I'm not sure of `Rational(1, 1)` strictly print as `1/1` , since the equation here is printed similar to the output of master except that brackets were added in the beginning of the expression.

Comment (oscarbenjamin) [reply sympy/printing/pretty/tests/test_pretty.py:1018]: `Rational(1, 1)` becomes `Integer(1)` which becomes `S.One`. There is no "unevaluated Rational" but you can do the same with a Pow if you want:
```python
In [72]: Pow(1, -1, evaluate=False)
Out[72]: 
1
─
1
```
